applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -136,15 +136,33 @@ File::File(Parameters Settings)
         m_file_id = H5Fcreate(m_file_name.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl_id);
     else
     {
-        if (file_driver != "core")
-            KRATOS_ERROR_IF(H5Fis_hdf5(m_file_name.c_str()) <= 0)
-                << "Invalid HDF5 file: " << m_file_name << std::endl;
+        /* Save old error handler */
+        herr_t (*old_func)(hid_t, void*);
+        void *old_client_data;
+
+        H5Eget_auto(H5E_DEFAULT, &old_func, &old_client_data);
+
+        /* Turn off error handling */
+        H5Eset_auto(H5E_DEFAULT, NULL, NULL);
+
+        /* Probe. Likely to fail, but that’s okay */
+        htri_t is_hdf5 = H5Fis_hdf5(m_file_name.c_str());
+
+        /* Restore previous error handler */
+        H5Eset_auto(H5E_DEFAULT, old_func, old_client_data);
+
         if (file_access_mode == "read_only")
         {
+            KRATOS_ERROR_IF(is_hdf5 <= 0 && file_driver != "core") << "Invalid HDF5 file: " << m_file_name << std::endl;
             m_file_id = H5Fopen(m_file_name.c_str(), H5F_ACC_RDONLY, fapl_id);
         }
-        else if (file_access_mode == "read_write")
-            m_file_id = H5Fopen(m_file_name.c_str(), H5F_ACC_RDWR, fapl_id);
+        else if (file_access_mode == "read_write") {
+            if (is_hdf5 <= 0) {
+                m_file_id = H5Fcreate(m_file_name.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl_id);
```

Comment:
@sunethwarna Does this imply that if the file does not exists, is firstly created here?

If so, there is something I don't understand, how the code knows when to create a new file because a new execution has begun and when to append the results to an existing file?

Reply:
it depends on the `file_access_mode`.

1. If the `file_access_mode` == 'truncate' : 
If existing file is found, the existing file will be truncated and new data is written. (This method will lose all the data in the existing file)
2. If the `file_access_mode` == 'exclusive' : 
If the exiting file is found, then an error is thrown. Otherwise the file will be created.
3. If the `file_access_mode` == 'read_only' : 
If the existing file is found, it is read, otherwise an error is thrown
4. If the `file_access_mode` == 'read_write' : 
If the existing file is found, then data will be overwritten (If the same file has same data groups and data sets, they will be overwritten in the sense of hdf5 overwrite). If the file is not found then file will be created.